Framework/Beamline/src/ComponentInfo.cpp

```diff
@@ -14,38 +15,179 @@ ComponentInfo::ComponentInfo()
+        throw std::invalid_argument("ComponentInfo should have been provided same "
+                                    "number of postions and rotations");
+      }
+      if (m_rotations->size() != m_detectorRanges->size()) {
+        throw std::invalid_argument(
+            "ComponentInfo should have as many positions "
+            "and rotations as assembly sorted detector component "
+            "ranges");
+      }
+      if (m_detectorInfo->size() != m_assemblySortedDetectorIndices->size()) {
+        throw std::invalid_argument("ComponentInfo must have detector indices "
+                                    "input of same size as size of DetectorInfo");
+      }
+      if (m_rotations->size() != m_componentRanges->size()) {
+        throw std::invalid_argument(
+            "ComponentInfo should have as many positions "
+            "and rotations as assembly sorted component "
+            "ranges");
+      }
+      if (m_assemblySortedComponentIndices->size() != m_size) {
+        throw std::invalid_argument("ComponentInfo must have component indices "
+                                    "input of same size as the sum of "
+                                    "non-detector and detector components");
+      }
+}
 
 std::vector<size_t>
 ComponentInfo::detectorIndices(const size_t componentIndex) const {
-  if (componentIndex < m_assemblySortedDetectorIndices->size()) {
+  if (isDetector(componentIndex)) {
     /* This is a single detector. Just return the corresponding index.
      * detectorIndex == componentIndex
      */
     return std::vector<size_t>{componentIndex};
   }
   // Calculate index into our ranges (non-detector) component items.
-  const auto rangesIndex =
-      componentIndex - m_assemblySortedDetectorIndices->size();
-  const auto range = (*m_ranges)[rangesIndex];
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  const auto range = (*m_detectorRanges)[rangesIndex];
   // Extract as a block
   return std::vector<size_t>(
       m_assemblySortedDetectorIndices->begin() + range.first,
       m_assemblySortedDetectorIndices->begin() + range.second);
 }
 
+std::vector<size_t>
+ComponentInfo::componentIndices(const size_t componentIndex) const {
+  if (isDetector(componentIndex)) {
+    /* This is a single detector. Just return the corresponding index.
+     * detectorIndex == componentIndex. Never any sub-components for detectors.
+     */
+    return std::vector<size_t>{componentIndex};
+  }
+  // Calculate index into our ranges (non-detector) component items.
+  const auto rangesIndex = compOffsetIndex(componentIndex);
+  const auto range = (*m_componentRanges)[rangesIndex];
+
+  // Extract as a block
+  return std::vector<size_t>(
+      m_assemblySortedComponentIndices->begin() + range.first,
+      m_assemblySortedComponentIndices->begin() + range.second);
+}
+
 size_t ComponentInfo::size() const { return m_size; }
 
 bool ComponentInfo::operator==(const ComponentInfo &other) const {
```

Comment:
I think these comparison operators should be removed now that we store positions/rotations.